### PR TITLE
fix: additional mailbox selector method on Props

### PR DESCRIPTION
**Pull request by Roiocam**
References #31862

Returning an empty mailbox is not a problem in PropsAdapter.

### CODE REVIEW

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSpec.scala [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+          Behaviors.stopped
+      }
+    }
+
+  "The Dispatcher selectors" must {
+
+    "default to unbounded" in {
+      val actor = spawn(behavior)
+      val (dispatcher, mailbox) = actor.ask(WhatsYourDispatcherAndMailbox.apply).futureValue
+      dispatcher.id eq Dispatchers.DefaultDispatcherId
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSpec.scala:54]:
```
dispatcher shouldBe Dispatchers.DefaultDispatcherId
```
?

Comment (patriknw) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSpec.scala:54]: and in more places below

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/PropsAdapter.scala [hunk 2]
```diff
@@ -23,13 +23,13 @@ import akka.dispatch.Mailboxes
 
   private final val TypedCreatorFunctionConsumerClazz = classOf[TypedCreatorFunctionConsumer]
   private final val ActorAdapterClazz = classOf[ActorAdapter[_]]
-  private final val DefaultTypedDeploy = Deploy.local.copy(mailbox = "akka.actor.typed.default-mailbox")
+  private final val DefaultTypedDeploy = Deploy.local
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/PropsAdapter.scala:26]: why is this changed? that mailbox is supposed to be the default
I think this would change the semantics of below
```
case _: DefaultMailboxSelector           => DefaultTypedDeploy
```

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSpec.scala (hunk 1 above)
Comment (Roiocam) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSpec.scala]: > ```
> dispatcher shouldBe Dispatchers.DefaultDispatcherId
> ```
> 
> ?

The types was mismatch, but I've already fixed it when constructing the return result.

File: akka-actor/src/main/scala/akka/dispatch/Mailboxes.scala [hunk 3]
```diff
@@ -182,10 +183,12 @@ private[akka] class Mailboxes(
       mailboxType
     }
 
-    if (deploy.mailbox != Deploy.NoMailboxGiven) {
+    if (deploy.mailbox != DefaultTypedMailboxId && deploy.mailbox != Deploy.NoMailboxGiven) {
```
Comment (Roiocam) [inline akka-actor/src/main/scala/akka/dispatch/Mailboxes.scala:186]: The mailbox selector priority is now as follows:

1. Custom mailbox
2. Custom dispatcher with mailbox configuration
3. Default typed actor mailbox
4. The originals...

By changing the order and using a redundant checks, Typed Actors and Classic Actors now has the same behavior on a custom dispatcher.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSelectorSpec.scala [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    "select an specific dispatcher from config with mailbox selector" in {
+      val actor =
+        spawn(behavior, DispatcherSelector.fromConfig("specific-dispatcher").withMailboxFromConfig("specific-mailbox"))
+      val (dispatcherId, mailbox) = actor.ask(WhatsYourDispatcherAndMailbox.apply).futureValue
+      mailbox shouldBe a[BoundedMessageQueueSemantics]
+      mailbox.asInstanceOf[BoundedNodeMessageQueue].capacity should ===(4)
+      dispatcherId eq "specific-dispatcher"
+    }
+
+    "select an specific dispatcher from config with mailbox config" in {
```
[the remark this answers is not in the document]
Comment (Roiocam) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSelectorSpec.scala:76]: It's reasonable that separate the approach of custom dispatcher and custom mailbox, still need a document commit mention this, existing typed akka document don't have any priority mailbox approache.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSelectorSpec.scala [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    "default to unbounded" in {
+      val actor = spawn(behavior)
+      val (_, mailbox) = actor.ask(WhatsYourDispatcherAndMailbox.apply).futureValue
+      mailbox shouldBe a[UnboundedMessageQueueSemantics]
+    }
+
+    "select a blocking dispatcher" in {
+      val actor = spawn(behavior, DispatcherSelector.blocking())
+      val (dispatcherId, _) = actor.ask(WhatsYourDispatcherAndMailbox.apply).futureValue
+      dispatcherId eq Dispatchers.DefaultBlockingDispatcherId
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSelectorSpec.scala:64]: `dispatcherId shouldBe` ?

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSelectorSpec.scala [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    "select an specific dispatcher from config with mailbox selector" in {
+      val actor =
+        spawn(behavior, DispatcherSelector.fromConfig("specific-dispatcher").withMailboxFromConfig("specific-mailbox"))
+      val (dispatcherId, mailbox) = actor.ask(WhatsYourDispatcherAndMailbox.apply).futureValue
+      mailbox shouldBe a[BoundedMessageQueueSemantics]
+      mailbox.asInstanceOf[BoundedNodeMessageQueue].capacity should ===(4)
+      dispatcherId eq "specific-dispatcher"
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSelectorSpec.scala:73]: `dispatcherId shouldBe` ?

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSelectorSpec.scala [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    "select an mailbox via dispatcher config would not work" in {
+      val actor = spawn(behavior, DispatcherSelector.fromConfig("specific-dispatcher-with-mailbox"))
+      val (dispatcherId, mailbox) = actor.ask(WhatsYourDispatcherAndMailbox.apply).futureValue
+      mailbox shouldBe a[NodeMessageQueue]
+      dispatcherId eq "specific-dispatcher-with-mailbox"
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSelectorSpec.scala:80]: `dispatcherId shouldBe` ?

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSelectorSpec.scala [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+          Behaviors.stopped
+      }
+    }
+
+  "The Dispatcher selectors" must {
+
+    "default to unbounded" in {
+      val actor = spawn(behavior)
+      val (_, mailbox) = actor.ask(WhatsYourDispatcherAndMailbox.apply).futureValue
+      mailbox shouldBe a[UnboundedMessageQueueSemantics]
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSelectorSpec.scala:58]: should this be even more specific?
```
mailbox shouldBe a[NodeMessageQueue]
```

**Review by patriknw — COMMENTED** [review]
looking good, just a few minor comments in the test

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSelectorSpec.scala (hunk 8 above)
Comment (Roiocam) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/DispatcherSelectorSpec.scala:58]: Thanks for the suggestion. However, when I received the CI complaint, I noticed that there is already a DispatcherSelectorSpec in akka.actor.typed.scaladsl. So, I want to figure out how to merge them. I will fix those issues and commit later.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/DispatcherSelectorSpec.scala [hunk 9]
```diff
@@ -8,119 +8,156 @@ import com.typesafe.config.Config
...
-      val grandParent = spawn(SpawnProtocol(), Props.empty.withDispatcherFromConfig("ping-pong-dispatcher"))
-      val parentProbe = createTestProbe[ActorRef[SpawnProtocol.Spawn[Ping]]]()
-      grandParent ! SpawnProtocol.Spawn(
-        SpawnProtocol(),
-        "parent",
-        Props.empty.withDispatcherSameAsParent,
-        parentProbe.ref)
-
-      val childProbe = createTestProbe[ActorRef[Ping]]()
-      grandParent ! SpawnProtocol.Spawn(PingPong(), "child", Props.empty.withDispatcherSameAsParent, childProbe.ref)
```
Comment (Roiocam) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/DispatcherSelectorSpec.scala:97]: It seems that the 'parent' and 'child' do not spawn with a vertical hierarchy, but rather a horizontal hierarchy.

Comment (patriknw) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/DispatcherSelectorSpec.scala:97]: yes, seems to be a mistake, it's spawning from grandParent twice

**Review by patriknw — APPROVED** [review]
LGTM, thanks

**Review by johanandren — APPROVED** [review]
LGTM, thanks!
